### DISCUSSION TITLE
从旧版本迁移中遇到的一些问题

### DISCUSSION THREAD

**Discussion by H4M5TER** [koishijs/koishi#144] [Q&A] [answered]
我之前使用koishi v1，通过node调用koishi，今天尝试迁移到koishi v3，遇到了一些问题。

1.创建App的参数问题
1.1为什么协议和id是Bot分别持有的，而监听端口是整个App持有的。
1.2无论是在快速上手页面还是在API的App页面都没有看到服务器地址参数，并且VSC也提示“server”不在类型“AppOptions”中。
1.3在API的App页面看到，协议的可选项是onebot和kaiheila，但是在快速上手页面看到，协议默认填写的是onebot:http。
2.尝试将app.sender.getGroupInfo改为session.$bot.getGroup，VSC提示类型“Bot”上不存在属性“getGroup”。
3.尝试将app.sender.sendGroupMsg改为app.bots[selfId].sendGroupMsg，此处的selfId应该是什么类型？，如果我只使用一个机器人，是否可以这样写：app.bots[0]？

**Comment by H4M5TER** [comment]
@Shigma

**Comment by shigma** [comment, marked as answer]
### 1.1 为什么协议和 id 是 Bot 分别持有的，而监听端口是整个 App 持有的？

首先协议和 id 是每个 bot 分别持有的是很自然的一件事。监听端口作为 App 的配置项是出于性能考虑。开放尽可能少的端口，合并尽可能多的服务能够提高资源的利用率。Koishi 内置了 Koa 作为服务端框架，通过 Router 可以有效实现内容的分流。

如果您需要开放多个端口，可以单独开一个 feat request，描述您的使用场景，我会考虑加入对应的支持。

**Reply by H4M5TER** [reply]
我是没有这方面的需求啦，不过按道理说会有这几种情况：

1. 多个机器人不同协议相同端口，不同协议的数据同时往一个端口里上报
2. 多个机器人相同协议相同端口，同一个协议的数据同时往一个端口里上报

不同协议的数据包或许可以理清楚，但是后者这种情况应该会发生混淆吧，所以感觉监听端口应该也是bot分别持有的才对？

**Reply by shigma** [reply]
@H4M5TER 我觉得问题正好相反

1. 如果是不同协议，建议配置不同的路由，每个协议通常都有自己的 `path` 配置项（例如 `options.onebot.path`）
2. 如果是相同协议，往往可以直接通过数据包识别机器人来源，至少目前支持的四个平台都提供了上报消息的标识，都可以用来区分是哪个号接收的

**Reply by H4M5TER** [reply]
了解了

**Comment by shigma** [comment]
### 1.2 无论是在快速上手页面还是在 API 的 App 页面都没有看到服务器地址参数，并且 VSC 也提示“server”不在类型“AppOptions”中？

因为 `server` 字段是 koishi-adapter-onebot 声明的，并不是所有平台都有的（事实上目前只有 onebot 一个平台需要这个字段）。如果您需要获得良好的代码提示，需要手动载入这个依赖：

```ts
import 'koishi-adapter-onebot'
new App({ server: 'http://localhost:5700' })
```

关于「快速上手」章节，我承认目前做的还不太好。我也正在摸索如何通过文档更有效地传达信息，如果您有什么建议欢迎直接在下面告诉我。

**Comment by shigma** [comment]
### 1.3 在 API 的 App 页面看到，协议的可选项是 onebot 和 kaiheila，但是在快速上手页面看到，协议默认填写的是 onebot:http。

只要您的配置正确，无论将 type 配置成 onebot 还是 onebot:http 都是可以运行的。这是因为 Koishi v3 使用了 [**适配器重定向**](https://koishi.js.org/guide/adapter.html#适配器重定向) 技术，它能够自动根据您的配置选择所需的协议。或许在「快速上手」一章中，使用简单的 onebot 更合适。

**Comment by shigma** [comment]
### 2 尝试将 `app.sender.getGroupInfo` 改为 `session.$bot.getGroup`，VSC提示类型“Bot”上不存在属性“getGroup”。

请参考最新的迁移文档，使用 `session.bot.getGroup`。

**Comment by shigma** [comment]
### 3 尝试将 `app.sender.sendGroupMsg` 改为 `app.bots[selfId].sendGroupMsg`，此处的 `selfId` 应该是什么类型？如果我只使用一个机器人，是否可以这样写：`app.bots[0]`？

第一个问题的答案是 `{platform}:{selfId}`，例如 `onebot:123456789`。

第二个问题的答案是可以。甚至我觉得这样写更简单。

**Comment by idlist** [comment]
关于3，在Koishi的实际使用中其实很少会app.bots[]，而是通过session来识别当前会话的信息，并且使用对应的方法。

**Reply by H4M5TER** [reply]
这个我了解，但是确实有接触不到上下文的地方，举例来说，我之前做的是b站动态的qq群推送

**Reply by shigma** [reply]
@H4M5TER 主动推送可以使用 [`ctx.broadcast()`](http://localhost:8080/api/context.html#ctx-broadcast)

**Reply by H4M5TER** [reply]
@Shigma 我检索了一下文档，似乎是bot.broadcast()？在接触不到上下文的地方依然需要app.bots[0]，不过这个接口对我有帮助，我之前是遍历数组挨个sendGroupMsg()的

**Reply by shigma** [reply]
@H4M5TER ctx 和 bot 都有 broadcast，功能不一样（

不可能接触不到上下文的，app 就是一个 ctx
